inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp

```diff
@@ -583,167 +197,107 @@ bool is_cv_type_in_list(const int type_id) {
-
-        const auto rowFunc = type_dispatch<merge_supported_types>(out.meta().depth, cv_type_id{}, typed_merge_row<2>{}, nullptr);
-        for (int l = 0; l < out.lpi(); l++) {
-            rowFunc({a.InLineB(l), b.InLineB(l)}, out.OutLineB(l), a.length());
+template<typename T, int chs>
+void mergeRowImpl(scalar_tag, const std::array<const T*, chs>& ins, T* out, const int length) {
+    for (int x = 0; x < length; ++x) {
+        for (int c = 0; c < chs; ++c) {
+            out[chs * x + c] = ins[c][x];
         }
     }
-};
+}
 
-GAPI_FLUID_KERNEL(FMerge3, Merge3, false) {
-    static const int LPI = 4;
-    static const int Window = 1;
-    static void run(const cv::gapi::fluid::View& a,
-                    const cv::gapi::fluid::View& b,
-                    const cv::gapi::fluid::View& c,
-                          cv::gapi::fluid::Buffer& out) {
-        GAPI_DbgAssert(is_cv_type_in_list<merge_supported_types>(out.meta().depth));
+template<typename isa_tag_t, int chs>
+struct typed_merge_row {
+    using p_f = void (*)(const std::array<const uint8_t*, chs>& ins, uint8_t* out, const int length);
 
-        const auto rowFunc = type_dispatch<merge_supported_types>(out.meta().depth, cv_type_id{}, typed_merge_row<3>{}, nullptr);
-        for (int l = 0; l < out.lpi(); l++) {
-            rowFunc({a.InLineB(l), b.InLineB(l), c.InLineB(l)}, out.OutLineB(l), a.length());
-        }
+    template <typename type>
+    typename std::enable_if<std::is_same<isa_tag_t, scalar_tag>::value ||
+            (!std::is_same<isa_tag_t, scalar_tag>::value && !std::is_same<type, uint8_t>::value &&
+             !std::is_same<type, float>::value), p_f>::type
+    operator()(type_to_type<type> ) {
+        return [](const std::array<const uint8_t*, chs>& ins, uint8_t* out, const int length) {
+            const auto inT = reinterpret_cast<const std::array<const type*, chs>&>(ins);
+            auto outT = reinterpret_cast<type*>(out);
+            scalar_tag t;
+            mergeRowImpl<type, chs>(t, inT, outT, length);
+        };
     }
-};
 
-GAPI_FLUID_KERNEL(FMerge4, Merge4, false) {
-    static const int LPI = 4;
-    static const int Window = 1;
-    static void run(const cv::gapi::fluid::View& a,
-                    const cv::gapi::fluid::View& b,
-                    const cv::gapi::fluid::View& c,
-                    const cv::gapi::fluid::View& d,
-                          cv::gapi::fluid::Buffer& out) {
-        GAPI_DbgAssert(is_cv_type_in_list<merge_supported_types>(out.meta().depth));
-
-        const auto rowFunc = type_dispatch<merge_supported_types>(out.meta().depth, cv_type_id{}, typed_merge_row<4>{}, nullptr);
-        for (int l = 0; l < out.lpi(); l++) {
-            rowFunc({a.InLineB(l), b.InLineB(l), c.InLineB(l), d.InLineB(l)}, out.OutLineB(l), a.length());
-        }
+    template<typename tag = isa_tag_t>
+    typename std::enable_if<!std::is_same<tag, scalar_tag>::value, p_f>::type
+    operator()(type_to_type<uint8_t>) {
```

Comment:
do we need `template tag` for method still instead of initial struct template?

Reply:
and if it is possible to eliminate method template tag and reuse struct templated tag here, then it seems that templated class `typed_merge_row ` has method which depends from class definition and seems it is more typical and easy to simply define class partial specialization for `scalar_tag` only as a classical and more easy-to-understand OOP solution

Reply:
C++11 and higher provides a lot of instruments diversity and gives a lot of reasons for over-engineering :)

Reply:
> do we need `template tag` for method still instead of initial struct template?

Yes, it necessary because of the specific syntax construction of the std::enable_if

Reply:
> and if it is possible to eliminate method template tag and reuse struct templated tag here, then it seems that templated class `typed_merge_row ` has method which depends from class definition and seems it is more typical and easy to simply define class partial specialization for `scalar_tag` only as a classical and more easy-to-understand OOP solution

This approach doesn't work because of the specific syntax construction of the std::enable_if